src/librustc_typeck/check/generator_interior.rs

```diff
@@ -0,0 +1,115 @@
+            if self.fcx.tcx.sess.verbose() {
+                if let Some(s) = scope {
+                    self.fcx.tcx.sess.span_warn(s.span(&self.fcx.tcx.hir).unwrap_or(DUMMY_SP),
+                        &format!("type in generator with scope = {:?}, type = {:?}",
+                                 scope,
+                                 self.fcx.resolve_type_vars_if_possible(&ty)));
+                } else {
+                    self.fcx.tcx.sess.span_warn(DUMMY_SP,
+                        &format!("type in generator WITHOUT scope, type = {:?}",
+                                 self.fcx.resolve_type_vars_if_possible(&ty)));
+                }
+                if let Some(e) = expr {
+                    self.fcx.tcx.sess.span_warn(e.span,
+                        &format!("type from expression: {:?}", e));
+                }
+            }
+            self.types.insert(ty);
+        } else if self.fcx.tcx.sess.verbose() {
+            if let Some(e) = expr {
+                self.fcx.tcx.sess.span_warn(e.span,
+                    &format!("NO type from expression: {:?}", e));
+            }
+        }
+    }
+}
+
+pub fn find_interior<'a, 'gcx, 'tcx>(fcx: &'a FnCtxt<'a, 'gcx, 'tcx>,
+                                     def_id: DefId,
+                                     body_id: hir::BodyId,
+                                     witness: Ty<'tcx>) {
+    let body = fcx.tcx.hir.body(body_id);
+    let mut visitor = InteriorVisitor {
+        fcx,
+        types: FxHashSet(),
+        region_maps: fcx.tcx.region_maps(def_id),
+    };
+    intravisit::walk_body(&mut visitor, body);
+
+    // Deduplicate types
+    let set: FxHashSet<_> = visitor.types.into_iter()
+        .map(|t| fcx.resolve_type_vars_if_possible(&t))
+        .collect();
+    let types: Vec<_> = set.into_iter().collect();
+
+    let tuple = fcx.tcx.intern_tup(&types, false);
+
+    if fcx.tcx.sess.verbose() {
+        fcx.tcx.sess.span_warn(body.value.span,
+            &format!("Types in generator {:?}", tuple));
+    }
+
+    // Unify the tuple with the witness
+    match fcx.at(&fcx.misc(body.value.span), fcx.param_env).eq(witness, tuple) {
+        Ok(ok) => fcx.register_infer_ok_obligations(ok),
+        _ => bug!(),
+   }
+}
+
+impl<'a, 'gcx, 'tcx> Visitor<'tcx> for InteriorVisitor<'a, 'gcx, 'tcx> {
+    fn nested_visit_map<'this>(&'this mut self) -> NestedVisitorMap<'this, 'tcx> {
+        NestedVisitorMap::None
+    }
+
+    fn visit_body(&mut self, _body: &'tcx Body) {
+        // Closures inside are not considered part of the generator interior
+    }
+
+    fn visit_pat(&mut self, pat: &'tcx Pat) {
+        if let PatKind::Binding(..) = pat.node {
+            let scope = self.region_maps.var_scope(pat.id);
+            let ty = self.fcx.tables.borrow().pat_ty(pat);
+            self.record(ty, Some(scope), None);
+        }
+
+        intravisit::walk_pat(self, pat);
+    }
+
+    fn visit_expr(&mut self, expr: &'tcx Expr) {
```

Comment:
This could also be done in `FnCtxt::write_ty` - it would need another argument, but there aren't that many call-sites, and you won't need to do another pass on the HIR.